maggy/optimizer/bayes/base.py

```diff
@@ -0,0 +1,1016 @@
+    of a trial.
+    Using additionally interim metrics of trials, i.e. use metrics that were trained for varying budgets makes it possible
+    to use early stopping in BO. (Also it makes it possible to have only one model in multi fidelity bo instead of one
+    model per fidelity)
+    For each interim metric a hyperparameter config augumented with the budget the metric was generated with is added
+    to the training data of the surrogate model
+
+    .. math::
+                                    z_t = [x_t, n_t] ; y_t = y_{t,nt}
+
+    When maximizing the acquisition function to sample the next hyperparameter config to evaluate, always augument with
+    the max budget N
+
+    .. math::
+                                    xt ← argmax acq([x, N])
+
+
+    """
+
+    def __init__(
+        self,
+        num_warmup_trials=15,
+        random_fraction=0.33,
+        interim_results=False,
+        interim_results_interval=10,
+        pruner=None,
+        pruner_kwargs=None,
+    ):
+        """
+        :param num_warmup_trials: number of random trials at the beginning of experiment
+        :type num_warmup_trials: int
+        :param random_fraction: fraction of random samples, between [0,1]
+        :type random_fraction: float
+        :param interim_results: If True, use interim metrics from trials for fitting surrogate model. Else use final
+                                metrics only
+        :type interim_results: bool
+        :param interim_results_interval: Specifies which interim metrics are used (if interim_results==True)
+                                         e.g. interval=10: the metric of every 10th epoch is used for fitting surrogate
+        :type interim_results_interval: int
+        :param pruner: name of pruning algorithm to use. So far only `hyperband` supported
+        :type pruner: str
+        :param pruner_kwargs: dict of arguments for initializing pruner. See pruner classes for reference.
+        :type pruner_kwargs: dict
+
+        Attributes
+        ----------
+
+        models (dict): The surrogate models for different budgets are saved in the `models` dict with the budgets as key.
+                       In case of a single fidelity optimization without a pruner. The only model has the key `0`.
+        warm_up_configs(list[dict]): list of hparam configs used for warming up
+        sampling_time_start (float): helper variable to calculate time needed for calculating next suggestion, i.e
+                                       when sampling from model (`sample_type`=="model"). Calulating the time happens
+                                       in `create_trial()`
+        normalize_categorical (bool): If True, the encoded categorical hparam is also max-min
+                                      normalized between 0 and 1 in searchspace.transform()
+        """
+        super().__init__()
+
+        # configure pruner
```

Comment:
I guess we can move this to `AbstractOptimizer` to deduplicate with `RandomSearch`

Reply:
yes